File: services/processors/SaqueProcessor.js

const ITransacaoProcessor = require('../../interfaces/ITransacaoProcessor');

class SaqueProcessor extends ITransacaoProcessor {
    constructor(contaRepository, logger) {
        super();
        this.contaRepository = contaRepository;
        this.logger = logger;
    }

    processar(transacao) {
        const contaOrigem = this.contaRepository.getSaldo(transacao.contaOrigem);
        
        if (!contaOrigem) {
            this.logger.log(`Conta ${transacao.contaOrigem} não encontrada para saque`);
            return false;
        }

        if (contaOrigem.saldo < transacao.valor) {
            this.logger.log(`Saque ${transacao.correlationId} cancelado por saldo insuficiente`);
            return false;
        }

        contaOrigem.saldo -= transacao.valor;
        this.contaRepository.atualizarSaldo(contaOrigem);

        this.logger.log(`Saque ${transacao.correlationId} processado. Novo saldo: ${contaOrigem.saldo}`);
        return true;
    }
}

module.exports = SaqueProcessor;

File: services/processors/saque_processor.py

from interfaces.i_transacao_processor import ITransacaoProcessor

class SaqueProcessor(ITransacaoProcessor):
    def __init__(self, conta_repository, logger):
        self.conta_repository = conta_repository
        self.logger = logger

    def processar(self, transacao):
        conta_saldo_origem = self.conta_repository.get_saldo(transacao.conta_origem)
        
        if not conta_saldo_origem:
            self.logger.log(f"Conta origem {transacao.conta_origem} não encontrada")
            return False
        
        if conta_saldo_origem.saldo < transacao.valor:
            self.logger.log(f"Saque numero {transacao.correlation_id} foi cancelado por falta de saldo")
            return False

        conta_saldo_origem.saldo -= transacao.valor
        self.conta_repository.atualizar_saldo(conta_saldo_origem)

        self.logger.log(f"Saque numero {transacao.correlation_id} foi efetivado com sucesso! Novo saldo: {conta_saldo_origem.saldo}")
        return True
